### PR TITLE
Magpietts 2503 s multi lingual

### PR DESCRIPTION
**Pull request by shehzeen**
* Updates Multilingual yaml to match the English yaml hyperparams
* Minor updates to support IPA chars in text field - computes CER/WER against a different text key in the manifest if available 'original_text'
* Bug fix in Whisper transcribe function for multilingual GRPO
* GRPO updates - added DR. GRPO loss function also

### CODE REVIEW

File: nemo/collections/tts/data/text_to_speech_dataset.py [hunk 1]
```diff
@@ -606,7 +606,12 @@ def __getitem__(self, index):
             align_prior = torch.tensor(align_prior, dtype=torch.float32)
             example["align_prior"] = align_prior
 
-        example['raw_text'] = data.text
+        if "original_text" in data.manifest_entry:
+            # For Manifests in which text field is IPA.
+            example['raw_text'] = data.manifest_entry['original_text']
+        else:
+            example['raw_text'] = data.text
```
Comment (blisc) [inline nemo/collections/tts/data/text_to_speech_dataset.py]: We should refactor this since this is a non-standard change to our manifest structure and dataset class

File: nemo/collections/tts/models/magpietts_preference_optimization.py [hunk 2]
```diff
@@ -435,6 +435,11 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
             # import ipdb; ipdb.set_trace()
             assert HAVE_TORCHAUDIO, "torchaudio is required for PESQ reward"
             self.squim_objective_model = SQUIM_OBJECTIVE.get_model()
+        
+        self.loss_type = self.cfg.get('loss_type', 'grpo')
+        assert self.loss_type in ['grpo', 'dr_grpo'], f"Unknown loss type: {self.loss_type}"
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts_preference_optimization.py]:
```suggestion
        if self.loss_type not in ['grpo', 'dr_grpo']:
            raise ValueError(f"Received loss_type of {self.loss_type}, but the model only accepts one of ['grpo', 'dr_grpo']")
```
Try to avoid assert statements, and raise Exceptions instead. If possible, mention how the user can fix this error as well.

File: scripts/magpietts/README_magpie_po.md [hunk 3]
```diff
@@ -136,42 +136,65 @@ python scripts/magpietts/dpo/create_text_contextpairs.py \
```
Comment (blisc) [inline scripts/magpietts/README_magpie_po.md:136]: In addition to a readme, can we add a simple test to make sure that the online policy path works on our released checkpoints? This would mean that we have to support a path to allow for older checkpoints to work on our current branch. To address issues from other teams.

File: examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml [hunk 4]
```diff
@@ -14,7 +14,7 @@ val_ds_meta: ???
 
 model:
   model_type: "multi_encoder_context_tts" # single_encoder_sv_tts, multi_encoder_context_tts, decoder_context_tts or decoder_pretrain_synthesizer
-  use_text_conditioning_encoder: false # If true, distilbert will be used to encode context_text if provided.
+  use_text_conditioning_encoder: true # If true, distilbert will be used to encode context_text if provided.
```
Comment (blisc) [inline examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml:17]: Is there a reason to change this parameter?

File: examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml [hunk 5]
```diff
@@ -123,24 +124,27 @@ model:
...
       num_workers: 4
       drop_last: true
+      pin_memory: true
 
   validation_ds:
     dataset:
       _target_: nemo.collections.tts.data.text_to_speech_dataset.MagpieTTSDataset
       dataset_meta: ${val_ds_meta}
-      min_duration: 0.5
+      sample_rate: ${sample_rate}
```
Comment (blisc) [inline examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml]: We removed sample_rate from our configs, can you change this here too?

**Review by blisc — CHANGES_REQUESTED** [review]

File: nemo/collections/tts/data/text_to_speech_dataset.py (hunk 1 above)
Comment (paarthneekhara) [reply nemo/collections/tts/data/text_to_speech_dataset.py]: What is raw_text being used for?

File: nemo/collections/tts/models/magpietts.py [hunk 6]
```diff
@@ -293,7 +293,10 @@ def load_state_dict(self, state_dict, strict=True):
         if strict == False:
             super().load_state_dict(state_dict, strict=False)
         for name, child in self.named_children():
-            if name in ['_speaker_verification_model', '_codec_model']:
+            if name in ['_speaker_verification_model', '_codec_model', '_reference_model',
```
Comment (paarthneekhara) [inline nemo/collections/tts/models/magpietts.py:313]: Should we create a list or a set like self.ignore_modules_when_loading = [], and then update it in the individual model classes as and when needed?

File: scripts/magpietts/evaluate_generated_audio.py [hunk 7]
```diff
@@ -71,9 +71,10 @@ def process_text(input_text):
     return single_space_text
 
 def transcribe_with_whisper(whisper_model, whisper_processor, audio_path, language, device):
+    print("Transcribing with Whisper...")
```
Comment (paarthneekhara) [inline scripts/magpietts/evaluate_generated_audio.py]: Remove print

File: examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml (hunk 4 above)
Comment (shehzeen) [reply examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml:17]: 'use_text_conditioning_encoder' is true in magpietts_en.yaml and the 'use_text_conditioning_encoder' is overridden in all baseline english sub files as true.

Comment (blisc) [reply examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml:17]: Should these parameters exist in the inference yaml? Shouldn't we inherit them from the base checkpoint instead?

File: scripts/magpietts/evaluate_generated_audio.py (hunk 7 above)
Comment (blisc) [reply scripts/magpietts/evaluate_generated_audio.py]: Inside of scripts/, I care less about print(). Within the main nemo/ directory (aka the library portion of the repo, and what actually gets installed when you pip install), we require nemo.logging instead of print()

File: nemo/collections/tts/data/text_to_speech_dataset.py (hunk 1 above)
Comment (shehzeen) [reply nemo/collections/tts/data/text_to_speech_dataset.py]: @paarthneekhara  `raw_texts` and  is used in preference optimization for CER and WER computation in DPO data generation and GRPO. This is when `text` field contains the IPA string (used as input to the model), but we need the original text for computing CER and WER of the generated audio.

@blisc by refactoring do you mean I should add the original_text field in DatasetSample?

```
@dataclass
class DatasetSample:
    dataset_name: str
    manifest_entry: Dict[str, Any]
    audio_dir: Path
    feature_dir: Path
    text: str
    original_text: str = None # Used for CER/WER computation when text field contains phonemized/processed text
    speaker: str
    speaker_index: int = None
    tokenizer_names: List[str] = None
   ```

File: scripts/magpietts/README_magpie_po.md (hunk 3 above)
Comment (shehzeen) [reply scripts/magpietts/README_magpie_po.md:136]: I added a test. `tests/functional_tests/L2_TTS_Fast_dev_runs_Magpietts_OnlinePO.sh` 

It now works with the old multi-encoder checkpoint with overrides, but I had to add `update_ckpt` function in magpietts, to rename `t5_encoder` to `encoder` and `t5_decoder` to `decoder`, similar to what we do in infer_and_evaluate.py, so that it works correctly with the old checkpoint.

File: examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml (hunk 4 above)
Comment (shehzeen) [reply examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml:17]: I think, we can eventually get rid of the inference yamls, and just keep the training ones. We mainly need to remove train_ds and val_ds and add test_ds instead, which should be doable from the command line. We can make a separate PR for that. It will avoid unnecessary updates to all set of yamls and keep the defaults the same between training and inference.

File: examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml (hunk 5 above)
Comment (shehzeen) [reply examples/tts/conf/magpietts/magpietts_multilingual_v1.yaml]: Done.

File: nemo/collections/tts/models/magpietts.py (hunk 6 above)
Comment (shehzeen) [reply nemo/collections/tts/models/magpietts.py:313]: I think that will make it more difficult to track which keys are being ignored and where the list is being updated, especially if we update the ignore list in individual classes and the load_state_dict is only in the base class.

File: nemo/collections/tts/data/text_to_speech_dataset.py (hunk 1 above)
Comment (blisc) [reply nemo/collections/tts/data/text_to_speech_dataset.py]: > @paarthneekhara raw_texts and is used in preference optimization for CER and WER computation in DPO data generation and GRPO. This is when text field contains the IPA string (used as input to the model), but we need the original text for computing CER and WER of the generated audio.

This should be explained in our code. Ideally, our dataset or model would have this docstringed somewhere. For this PR, we should add this explanation here.

File: scripts/magpietts/README_magpie_po.md (hunk 3 above)
Comment (blisc) [reply scripts/magpietts/README_magpie_po.md:136]: Can you add this test to the CI suite? You need to add this script to github/workflows/cicd-main.yml . See https://github.com/NVIDIA/NeMo/pull/13831 for an example.

File: nemo/collections/tts/data/text_to_speech_dataset.py (hunk 1 above)
Comment (shehzeen) [reply nemo/collections/tts/data/text_to_speech_dataset.py]: Added comments about this.

File: scripts/magpietts/README_magpie_po.md (hunk 3 above)
Comment (shehzeen) [reply scripts/magpietts/README_magpie_po.md:136]: I missed that earlier. Thanks, I have added it now. It must be mentioned somewhere, but is there a way to run CI/CD or the tests locally before pushing?

Comment (blisc) [reply scripts/magpietts/README_magpie_po.md:136]: Can you also add a reference to your new test on line 1828 on .github/workflows/cicd-main.yml?
> but is there a way to run CI/CD or the tests locally before pushing?

Yes. However I haven't found an easy to do so yet, so not yet.

**Review by blisc — APPROVED** [review]
